tomlkit/items.py

```diff
@@ -1512,6 +1519,12 @@ def _new(self, result):
     def _getstate(self, protocol=3):
         return self._t, str(self), self._original, self._trivia
 
+    @classmethod
+    def from_raw(cls, value: str, type_=StringType.SLB, escape=False) -> "String":
+        string_value = escape_string(value) if escape else value
```

Comment:
`escape_string()` needs to take the string type as a parameter to decide whether a quote should be escaped or not.

Reply:
I was revisiting the standard and It seems that escaping is more complicated than that :/

I did some changes to the implementation of `escape_string` to be parameterised in terms of `escape_chars` and `escape_sequences`, and then I defined both groups for each string type.

Hopefully it works better now? Please let me know what you think!

Reply:
I'm sorry I may make a mistake here. According to toml's spec, there is no escaping in literal strings. Therefore, `'\t'(python) != '\t'(toml)`, and:
>Since there is no escaping, there is no way to write a single quote inside a literal string enclosed by single quotes.

If we are to build a string with `string("'", literal=True, multiline=False)`, it should raise an error.

Reply:
Yes: `'\t'(python) != '\t'(toml)`, this is correct. `'\t' (toml) == r'\t' (python).

So we have 2 alternatives here:

a) Raise an error when invalid characters are added with `literal=True`
b) Make the string valid... (i.e. be lenient with the input but strict with the output)

Since the `escape_string` function is not public, I implemented (b) because I think it is easier to use. If we consider that the user can still pass `escape=False`, this approach is the one with maximum flexibility.

I tried to describe it in the docstring. Hopefully is easier to understand?

Do you think that it is better to fail instead?

Reply:
I prefer solution a)

It is sometimes not possible to make the string valid when `literal=True`. So the whole logic would be:

1. `literal=False, multiline=False`. The escape characters are the MUST escapes plus `\t` and `"`(Yes, we prefer to escape `\t` for better looking)
2. `literal=False, multiline=True`. The escape sequences are the MUST escapes plus `\t` and `"""`
3. `literal=True, multiline=False`. No escapes, raise an error if `'` is in the string.
4. `literal=True, multiline=False`. No escapes, raise an error if `'''` is in the string.

Reply:
Thank you.

Just submitted a new patch with the method failing for invalid characters as discussed. Please let me know your thoughts.